src/ORM/DataObjectSchema.php

```diff
@@ -486,29 +500,65 @@ protected function cacheDatabaseFields($class)
     }
 
     /**
-     * Cache all indexes for the given class.
-     * Will do nothing if already cached
+     * Cache all indexes for the given class. Will do nothing if already cached.
      *
      * @param $class
      */
     protected function cacheDatabaseIndexes($class)
     {
-        if (array_key_exists($class, $this->databaseIndexes)) {
-            return;
+        if (!array_key_exists($class, $this->databaseIndexes)) {
+            $this->databaseIndexes[$class] = array_merge(
+                $this->defaultDatabaseIndexes($class),
+                $this->customDatabaseIndexes($class)
+            );
         }
+    }
+
+    /**
+     * Get "default" database indexable field types
+     *
+     * @param  string $class
+     * @return array
+     */
+    protected function defaultDatabaseIndexes($class)
+    {
         $indexes = [];
+        if (empty($this->defaultDatabaseIndexes[$class])) {
+            $this->defaultDatabaseIndexes[$class] = [];
+        }
 
-        // look for indexable field types
-        foreach ($this->databaseFields($class, false) as $field => $type) {
-            if ($type === 'ForeignKey' || $type === 'DBClassName') {
-                $indexes[$field] = [
-                    'type' => 'index',
-                    'columns' => [$field],
-                ];
+        // Ensure that fixed field types, e.g. ClassName can be indexed
+        $databaseFields = $this->databaseFields($class, false);
```

Comment:
If you use `$this->fieldSpecs($class, UNINHERITED)` here you would get all DBField instances, including DBPolymorphicForeignKey, DBForeignKey, and other instances.

All you need is the class of the field, and the `indexed` config option for that class right? then all you need is a method that gets the list of fields to index for a given DBField instance.